core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -147,46 +161,79 @@ public void readBytes(OutputStream dest, int length) throws IOException {
     execute(STREAM_OP, length, dest, 0);
   }
 
+  private static final NoThrowReadOperation<CompositeReadableBuffer> COMPOSITE_OP =
+      new NoThrowReadOperation<CompositeReadableBuffer>() {
+        @Override
+        public int read(ReadableBuffer buffer, int length, CompositeReadableBuffer dest,
+            int unused) {
+          dest.addBuffer(buffer.readBytes(length));
+          return 0;
+        }
+      };
+
   @Override
   public ReadableBuffer readBytes(int length) {
-    if (length <= 0) {
-      return ReadableBuffers.empty();
+    final CompositeReadableBuffer newBuffer = new CompositeReadableBuffer();
```

Comment:
Oops, I commented in https://github.com/grpc/grpc-java/pull/8102#discussion_r618687981

Reply:
@voidzcy returning the buffer directly is only one part of the optimization. You can still avoid allocating a composite buffer since `readBytes(length)` already allocates a new buffer with its own marker.

How about this:
```java
    ReadableBuffer newBuffer = null;
    CompositeReadableBuffer newComposite = null;
    do {
      ReadableBuffer buffer = buffers.peek();
      int readable = buffer.readableBytes();
      ReadableBuffer readBuffer;
      if (readable > length) {
        readBuffer = buffer.readBytes(length);
        length = 0;
      } else {
        if (marked) {
          readBuffer = buffer.readBytes(readable);
          advanceBufferIfNecessary(); // it _is_ necessary, so this will do couple of redundant checks
        } else {
          readBuffer = buffers.poll();
        }
        length -= readable;
      }
      if (newBuffer == null) {
        newBuffer = readBuffer;
      } else {
        if (newComposite == null) {
          newComposite =
              new CompositeReadableBuffer(length == 0 ? 2 : Math.min(buffers.size() + 2, 16));
          newComposite.addBuffer(newBuffer);
          newBuffer = newComposite;
        }
        newComposite.addBuffer(readBuffer);
      }
    } while (length > 0);
    return newBuffer;
```

Reply:
Restore the change. I previously thought that was too verbose compared to the amount of optimizations it can offer :P

Reply:
Thanks @voidzcy, agree that the other approach is much more elegant.